aiomisc/aggregate.py

```diff
@@ -0,0 +1,215 @@
+
+        if self.count == self.max_count:
+            event.set()
+            self._clear()
+        else:
+            # Waiting for max_count requests or a timeout
+            try:
+                await wait_for(
+                    event.wait(),
+                    timeout=first_call_at + self._leeway - monotonic(),
+                )
+            except asyncio.TimeoutError:
+                log.debug(
+                    'Aggregation timeout of %s for batch started at %.4f '
+                    'with %d calls after %.2f ms',
+                    self._func.__name__, first_call_at, len(futures),
+                    (monotonic() - first_call_at) * 1000,
+                )
+
+        # Clear only if not cleared already
+        if args is self._args:
+            self._clear()
+
+        # Trying to acquire the lock to execute the aggregated function
+        async with lock:
+            if not future.done():
+                await self._execute(args=args, futures=futures)
+        await future
+        return future.result()
+
+
+class AggregatorLowLevel(Aggregator):
+
+    async def _execute(self, *, args: list, futures: List[Future]) -> None:
+        args = [
+            Arg(value=arg, future=future)
+            for arg, future in zip(args, futures)
+        ]
+        try:
+            await self._func(*args)
+        except CancelledError:
+            # Other waiting tasks can try to finish the job instead.
+            raise
+        except Exception as e:
+            self._set_exception(e, futures)
+            return
+
+        # Validate that all results/exceptions are set by the func
+        for future in futures:
+            if not future.done():
+                future.set_exception(ResultNotSet)
+
+
+def aggregate(leeway_ms: float, max_count: int = None) -> Callable:
+    """
+    Parametric decorator that aggregates multiple
+    (but no more than ``max_count`` defaulting to ``None``) single-argument
+    executions (``res1 = await func(arg1)``, ``res2 = await func(arg2)``, ...)
+    of an asynchronous function with variadic positional arguments
+    (``async def func(*args, pho=1, bo=2) -> Iterable``) into its single
+    execution with multiple positional arguments
+    (``res1, res2, ... = await func(arg1, arg2, ...)``) collected within a time
+    window ``leeway_ms``. Note 1: ``func`` must return a sequence of values
+    of length equal to the number of arguments (and in the same order).
+    Note 2: if some unexpected error occurs, exception is propagated to each
+    future; to set an individual error for each aggregated call refer
+    to ``aggregate_ll``.
+    :param leeway_ms: The maximum approximate delay between the first
+    collected argument and the aggregated execution.
+    :param max_count: The maximum number of arguments to call decorated
+    function with. Default ``None``.
+    :return:
+    """
+    def _(func: AggFuncHighLevel) -> Callable[[Any], Awaitable]:
```

Comment:
please rename it as `wrapper` etc.

Reply:
Does not help with anything, just deteriorates readability.